### PR TITLE
Airbyte CDK: add CustomRecordFilter

### PR DESCRIPTION
**Pull request by artem1205**
## What

Resolving https://github.com/airbytehq/airbyte-internal-issues/issues/2919

## How

- add CustomRecordFilter

## Recommended reading order
1. `airbyte-cdk/python/airbyte_cdk/sources/declarative/declarative_component_schema.yaml`
2. `airbyte-cdk/python/airbyte_cdk/sources/declarative/models/declarative_component_schema.py`
## 🚨 User Impact 🚨

no breaking changes

## Pre-merge Actions
<details><summary><strong>Updating the Python CDK</strong></summary>

### Airbyter

Before merging:
- Pull Request description explains what problem it is solving
- Code change is unit tested
- Build and my-py check pass
- Smoke test the change on at least one affected connector
   - On Github: Run [this workflow](https://github.com/airbytehq/airbyte/actions/workflows/connectors_tests.yml), passing `--use-local-cdk --name=source-<connector>` as options
   - Locally: `airbyte-ci connectors --use-local-cdk --name=source-<connector> test`
- PR is reviewed and approved
      
After merging:
- [Publish the CDK](https://github.com/airbytehq/airbyte/actions/workflows/publish-cdk-command-manually.yml)
   - The CDK does not follow proper semantic versioning. Choose minor if this the change has significant user impact or is a breaking change. Choose patch otherwise.
   - Write a thoughtful changelog message so we know what was updated.
- Merge the platform PR that was auto-created for updating the Connector Builder's CDK version
   - This step is optional if the change does not affect the connector builder or declarative connectors.

</details>

### CODE REVIEW

**Comment by maxi297** [comment]
What is the filtering that we want to do for source-mailchimp? I'm asking because I feel like creating a custom component is something very different than adding the ability to create custom components where we couldn't before so I would prefer to add the capability in `RecordFilter` if trivial

**Comment by artem1205** [comment]
here are the components: https://github.com/airbytehq/airbyte/pull/35281/files#diff-a418efe8052a0a7074dac5278029a194c94ecec255f198ca83e19c991dac91c2R86-R117

**Comment by artem1205** [comment]
JInja2 interpolation does not support any "complex" expressions, e.g. list comprehensions, that's why we have to implement custom components

**Comment by natikgadzhi** [comment]
I'm so excited to see contributions to the CDK itself! I know it's not unique and you've been doing this for a while, but hey, I'm happy to see one.

How would it look like if we wanted to not just allow custom filter class ref declarations, but also bring more sophisticated record filtering into the low-code CDK and the Builder itself? Is there record filtering support in the low-code CDK at all? I've seen record field transforms, but not whole record filters I think?

**Comment by artem1205** [comment]
> How would it look like if we wanted to not just allow custom filter class ref declarations, but also bring more sophisticated record filtering into the low-code CDK and the Builder itself?

If we want to apply some custom transformation on the response (before we get the list of records), we need to implement it during the extraction. We have only DpathExtractor as of now.


> Is there record filtering support in the low-code CDK at all? 

we have a `RecordFilter` class, that takes as arguments list of records and jinja condition (boolean). this condition is applied for all incoming records and only condition=true are returned.

> I've seen record field transforms, but not whole record filters I think?

both record transofrmers `AddFields` and RemoveFields operate on record itself and return transformed record, e.g.
```python3
for record in records:
   transformation(record)
return records

```

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/declarative_component_schema.yaml [hunk 1]
```diff
@@ -516,6 +516,27 @@ definitions:
       $parameters:
         type: object
         additionalProperties: true
+  CustomRecordFilter:
+    title: Custom Record Extractor
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/declarative_component_schema.yaml:520]: this is a filter, not an extractor

**Review by girarda — APPROVED** [review]
small nit in the title and description. looks good otherwise! 🚢

**Comment by artem1205** [comment]
image

**Pull request merged**